### PR TITLE
✨ Add 7 workflow examples for MLJFlux

### PR DESCRIPTION
**Pull request by EssamWisam**
Added the following workflow examples:
image

Sorry if I didn't make the PR in the expected time. I encountered interruptions and was stuck in live training for longer than I wanted. In particular, I was trying hard to get it to work in Jupyter notebooks.

Will make another PR for a single tutorial soon.

### CODE REVIEW

**Comment by EssamWisam** [comment]
Added commit for another tutorial.

File: docs/src/full tutorials/Spam Detection with RNNs/SMS.jl [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+using WordTokenizers    # For tokenization
+using Languages         # For stop words
+
+
+# ### Reading Data
+df = CSV.read("./sms.csv", DataFrame);
+
+# Display the first 5 rows with DataFrames
+first(df, 5) |> pretty
+
```
Comment (ablaom) [inline docs/src/full tutorials/Spam Detection with RNNs/SMS.jl:25]: FYI. I used to use `pretty` a lot but maybe this is not really necessary as  DataFrames display is pretty sophisticated now.

File: docs/src/full tutorials/Spam Detection with RNNs/SMS.jl [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+# For the model, we will use a RNN from Flux. We will average the hidden states corresponding to any sequence then pass that to a dense layer for classification.
+
+# For this, we need to define a custom Flux layer to perform the averaging operation:
+
+struct Mean end
+Flux.@layer Mean
+(m::Mean)(x) = mean(x, dims = 2)[:, 1, :]   # [batch_size, seq_len, hidden_dim] => [batch_size, 1, hidden_dim]=> [batch_size, hidden_dim]
+
+# For compatibility, we will also define a layer that simply casts the input to integers as the embedding layer in Flux expects integets but the MLJFlux model expects floats:
```
Comment (ablaom) [inline docs/src/full tutorials/Spam Detection with RNNs/SMS.jl:123]:
```suggestion
# For compatibility, we will also define a layer that simply casts the input to integers as the embedding layer in Flux expects integers but the MLJFlux model expects floats:
```

Maybe do a spell check?

File: docs/src/full tutorials/Spam Detection with RNNs/SMS.jl [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+struct Intify end
+Flux.@layer Intify
+(m::Intify)(x) = Int.(x)
+
+# Here we define out network:
```
Comment (ablaom) [inline docs/src/full tutorials/Spam Detection with RNNs/SMS.jl:128]:
```suggestion
# Here we define our network:
```

File: docs/src/workflow examples/Basic Neural Architecture Search/Project.toml [hunk 4]
```diff
@@ -0,0 +1,7 @@
+[deps]
+DataFrames = "a93c6f00-e57d-5684-b7b6-d8193f3e46c0"
+Flux = "587475ba-b771-5e3f-ad9e-33799f191a9c"
+Literate = "98b081ad-f1c9-55d3-8b20-4c87d4299306"
```
Comment (ablaom) [inline docs/src/workflow examples/Basic Neural Architecture Search/Project.toml:4]: Not sure including `Literate` in the project is necessary or appropriate. See e.g. https://github.com/ablaom/MLJTutorial.jl/tree/dev/notebooks/01_data_representation .

File: docs/src/workflow examples/Basic Neural Architecture Search/tuning.jl [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    ## Define the range of neurons
+	neuron_range = min_neurons:neuron_step:max_neurons     
+
+    ## Empty list to store the network configurations
+	networks = Vector{Tuple{Vararg{Int, num_layers}}}()    
+
+	## Recursive helper function to generate all combinations of tuples
+	function generate_tuple(current_layers, remaining_layers)
+		if remaining_layers > 0
+			for n in neuron_range
```
Comment (ablaom) [inline docs/src/workflow examples/Basic Neural Architecture Search/tuning.jl:62]: nit picking: I think a indentation of 4 spaces is the Blue Style standard. Do not use TABs as Literate and Documenter don't seem to render them as expected.

File: docs/src/workflow examples/Basic Neural Architecture Search/tuning.jl [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+			## current_layers =[] then current_layers=[(min_neurons)], 
+                ## [(min_neurons+neuron_step)], [(min_neurons+2*neuron_step)],...
+			## for each of these we call generate_layers again which appends
+                ## the n combinations for each one of them
+				generate_tuple(vcat(current_layers, [n]), remaining_layers - 1)
```
Comment (ablaom) [inline docs/src/workflow examples/Basic Neural Architecture Search/tuning.jl:67]: Lines 63-66 don't appear to have consistent whitespace indentation.

File: docs/src/workflow examples/Basic Neural Architecture Search/tuning.jl [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+			end
+		else
+		## in the base case, no more layers to "recurse on" 
+            ## and we just append the current_layers as a tuple
+			push!(networks, tuple(current_layers...))
```
Comment (ablaom) [inline docs/src/workflow examples/Basic Neural Architecture Search/tuning.jl:72]: again in 70-71

**Comment by ablaom** [comment]
Since you propose a tutorial based on an RNN, you should probably study https://github.com/FluxML/MLJFlux.jl/issues/225. See in particular, the point about the observation dimension, near the end.

File: docs/src/workflow examples/Comparison/Project.toml [hunk 8]
```diff
@@ -0,0 +1,13 @@
+[deps]
+DataFrames = "a93c6f00-e57d-5684-b7b6-d8193f3e46c0"
+DecisionTree = "7806a523-6efd-50cb-b5f6-3fa6f1930dbb"
+Flux = "587475ba-b771-5e3f-ad9e-33799f191a9c"
+Literate = "98b081ad-f1c9-55d3-8b20-4c87d4299306"
```
Comment (ablaom) [inline docs/src/workflow examples/Comparison/Project.toml:5]: rm Literate. see previous comment.

File: docs/src/workflow examples/Composition/Project.toml [hunk 9]
```diff
@@ -0,0 +1,9 @@
+[deps]
+EarlyStopping = "792122b4-ca99-40de-a6bc-6742525f08b6"
+Flux = "587475ba-b771-5e3f-ad9e-33799f191a9c"
+Imbalance = "c709b415-507b-45b7-9a3d-1767c89fde68"
+Literate = "98b081ad-f1c9-55d3-8b20-4c87d4299306"
```
Comment (ablaom) [inline docs/src/workflow examples/Composition/Project.toml:5]: ditto

File: docs/src/workflow examples/Composition/composition.jl [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+Imbalance.checkbalance(y)        
+
+
+
+# ### Instantiating the model
+
+# Let's load `BorderlineSMOTE1` to oversample the data and `Standardizer` to standardize it.
+BorderlineSMOTE1 = @load BorderlineSMOTE1 pkg=Imbalance verbosity=0
+NeuralNetworkClassifier = @load NeuralNetworkClassifier pkg=MLJFlux
+Standardizer = @load Standardizer pkg=MLJModels
```
Comment (ablaom) [inline docs/src/workflow examples/Composition/composition.jl:39]: This is not incorrect, but `Standardizer` is always available without needing to explicitly load it. See all built-in models by doing `local models()` at startup.

File: docs/src/workflow examples/Composition/composition.jl [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+# Now let's compose the balanced model with a standardizer.
+pipeline = standarizer |> balanced_model
+# By this, any training data will be standardized then oversampled then passed to the model. Meanwhile,
+# for inference, the standardizer will automatically use the training set's mean and std and the oversampler
+# will be transparent.
+
+
+# ### Training the Composed Model
+# It's indistinguishable from training a single model. Isn't MLJ beautiful?
```
Comment (ablaom) [inline docs/src/workflow examples/Composition/composition.jl:63]: Appreciate the enthusiasm, but let's drop that last sentence 🤫 

```suggestion
# It's indistinguishable from training a single model.
```

**Comment by ablaom** [comment]
Looks like the Early Stopping markdown did not generate (I'm seeing "empty file").

File: docs/src/workflow examples/Early Stopping/iteration.jl [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+]
+
+# Construct the iterated model and pass to it the stop_conditions and the callbacks:
+iterated_model = IteratedModel(model=clf,
+                               resampling=CV(nfolds=6),    # Split the data internally into 0.7 training and 0.3 validation
+                               measures=log_loss,
+                               iteration_parameter=:(epochs),
+                               controls=vcat(stop_conditions, callbacks),
+                               retrain=false                  # no need to retrain on all data at the end
+                               );
```
Comment (ablaom) [inline docs/src/workflow examples/Early Stopping/iteration.jl:62]: Whitespace suggestion:

```suggestion
iterated_model = IteratedModel(
    model=clf,
    resampling=CV(nfolds=6),    # Split the data internally into 0.7 training and 0.3 validation
    measures=log_loss,
    iteration_parameter=:(epochs),
    controls=vcat(stop_conditions, callbacks),
    retrain=false                  # no need to retrain on all data at the end
);
```

File: docs/src/workflow examples/Hyperparameter Tuning/Project.toml [hunk 13]
```diff
@@ -0,0 +1,7 @@
+[deps]
+Flux = "587475ba-b771-5e3f-ad9e-33799f191a9c"
+Literate = "98b081ad-f1c9-55d3-8b20-4c87d4299306"
```
Comment (ablaom) [inline docs/src/workflow examples/Hyperparameter Tuning/Project.toml:3]: ditto

File: docs/src/workflow examples/Incremental Training/Project.toml [hunk 14]
```diff
@@ -0,0 +1,6 @@
+[deps]
+Flux = "587475ba-b771-5e3f-ad9e-33799f191a9c"
+Literate = "98b081ad-f1c9-55d3-8b20-4c87d4299306"
```
Comment (ablaom) [inline docs/src/workflow examples/Incremental Training/Project.toml:3]: ditto

File: docs/src/workflow examples/Live Training/Project.toml [hunk 15]
```diff
@@ -0,0 +1,7 @@
+[deps]
+Flux = "587475ba-b771-5e3f-ad9e-33799f191a9c"
+Literate = "98b081ad-f1c9-55d3-8b20-4c87d4299306"
```
Comment (ablaom) [inline docs/src/workflow examples/Live Training/Project.toml:3]: ditto

File: docs/src/workflow examples/Live Training/live-training.jl [hunk 16]
```diff
@@ -0,0 +1,73 @@
+# # Incremental Training with MLJFlux
```
Comment (ablaom) [inline docs/src/workflow examples/Live Training/live-training.jl:1]:
```suggestion
# # Live Training with MLJFlux
```

**Review by ablaom — COMMENTED** [review]
Herculean effort, thanks!

(I haven't checked the notebooks or markdown).

**Comment by ablaom** [comment]
After some more thought, I wonder if the spam example isn't little too long and challenging for documentation.

You're representing messages as sequences of integers - one integer per word in some fixed vocabulary (and a special integer for words outside). To train an RNN-based classifier you need to one-hot encode but to do that "lazily" you want to use Flux's embedding layer, rather than an external one-hot encoding transformer which will just produce a very large table. However, an obstacle in your approach is that is the MLJFlux's NeuralNetworkClassifier does not support categorical data, so you have to fool it by representing integers as floats, and then reverse this in a pre-processing layer of the NN preceding the Embed layer. Do I have this right?

I can see how all this should work, and it's great to see this solution, however it's definitely not MLJ idiomatic, which adds a cognitive burden for users who want to first understand the "proper" way of doing things. 

Of course, after the summer project, you will have extended MLJFlux to overcome this kind of issue, but with something probably more useful than one-hot encoding  for high cardinality features (we'll be learning an embedding directly into a lower dimensional space) and something that can be transferred to other classifiers. So maybe a new version of this example would be appropriate then.

File: docs/src/full tutorials/Spam Detection with RNNs/SMS.jl (hunk 1 above)
Comment (EssamWisam) [reply docs/src/full tutorials/Spam Detection with RNNs/SMS.jl:25]: Agree. I like DF display more.

File: docs/src/full tutorials/Spam Detection with RNNs/SMS.jl (hunk 2 above)
Comment (EssamWisam) [reply docs/src/full tutorials/Spam Detection with RNNs/SMS.jl:123]: OK.

File: docs/src/workflow examples/Basic Neural Architecture Search/Project.toml (hunk 4 above)
Comment (EssamWisam) [reply docs/src/workflow examples/Basic Neural Architecture Search/Project.toml:4]: I did not know that one does not need to `add Literate` to use it. This is what I understood from what you said and after performing a quick experiment.

So far, two actions I need to be perform globally over the tutorials:
- Spell checking
- Ditching `Literate` from `Project.toml`

File: docs/src/workflow examples/Basic Neural Architecture Search/tuning.jl (hunk 5 above)
Comment (EssamWisam) [reply docs/src/workflow examples/Basic Neural Architecture Search/tuning.jl:62]: Will try to see if I can tell my Julia formatter to use spaces. I noticed that it incorrectly renders comments which is why I made the spacing non uniform for the comments.

File: docs/src/workflow examples/Composition/composition.jl (hunk 10 above)
Comment (EssamWisam) [reply docs/src/workflow examples/Composition/composition.jl:39]: OK. I will add a comment next to it to clarify it's not necessary.

**Comment by EssamWisam** [comment]
> Looks like the Early Stopping markdown did not generate (I'm seeing "empty file").

@ablaom for some reason, I can see the full files in the Github [review log](https://github.com/FluxML/MLJFlux.jl/pull/256/files/975ed6e72a4a25bad9ae5fdafd6be1bb7759b4a3#diff-bdff67f24705329aa1895d57f5ea50bf5feacf52e486c16928bcb6380dfa8661).

**Comment by EssamWisam** [comment]
> After some more thought, I wonder if the spam example isn't little too long and challenging for documentation.

It's specifically in the tutorials section of the documentation which I think doesn't have to be limited to less challenging tutorials (esp. with the rest of the docs and the workflow examples being present).

As for it being potentially long, I think it's shorter than Boston and MNIST  tutorials and I tried to be systematic in explaining it so one does not get lost. As for it being potentially challenging, I would say unlikely unless for someone that did not come accross using RNNs before (in which case it helps them learn more about RNNs but isn't so easy because if I were to explain more on this it would feel like a lecture and be quite long for those that have came across using them and just want to see more applications; also tokenization, embedding and other concepts are not missed in NLP courses).

> This is my understanding: You're representing messages as sequences of integers ... To train an RNN-based classifier you need to one-hot encode but to do that lazily.

Yes, the embedding layer simulates the functionality that will be performed on one-hot vectors with an equivalent functionality over integers which makes it more efficient. 

> An obstacle in your approach is that is the MLJFlux's NeuralNetworkClassifier does not support categorical data, so you have to fool it by representing integers as floats, and then reverse this in a pre-processing layer of the NN preceding the Embed layer. Do I have this right?

Yes, correct. A sentence in this case is viewed as a sequence of categorical columns where each takes values from the same set (the vocabulary).

> I can see how all this should work, and it's great to see this solution, however it's definitely not MLJ idiomatic, which adds a cognitive burden for users who want to first understand the "proper" way of doing things.

Yes, I agree, but if we consider the degree of deviation from MLJ-idiomatic code we find it's only the `Intify` layer which is certainly easy to understand and is well-justified (i.e., has neglible effects on learning or generalizing to other applications because it should be well clear why we added this very minor layer). 

Minor comment on "this should work", yes it has expectedly worked as in the executed notebook/markdown.

> Of course, after the summer project, you will have extended MLJFlux to overcome this kind of issue...

After some reflection, I would like to emphasize that what follows the three dots may not be exactly precise. In particular, the difference from MLJFlux's embedding layer and the custom layer that will be added to deal with categorical data is that the former will share weights among all words (and that's useful); meanwhile, the latter assumes each column in the vector is another categorical variables with different categories and by default uses an embedding layer for each column.

Thereby, I have thought about two resolutions for this issue:

**Resolution A**:
- Allow `Count` scientific type aside from floats in MLJFlux classifier/regressor (can be done now)
- The embedding layer that will be added in the summer project will deal with `Multiclass`/`Ordinal` columns (and will ignore `Count`/`Float` columns.

Why **A** makes sense:
- The tutorial will immediately become MLJ idiomatic (X coerced to integers and no need for `Intify`)
- In general, my classification or regression data may have integers that are still not categorical
- The new custom embedding layer for categorical data will be ignored in cases like this (all data is Counts) which is what we need as one embedding layer only should be used (and not a different embedding layer for each index in the sequence).

**Resolution B**:
- Still allow `Count` scientific types for the aforementioned reason (my classification or regression data may have integers that are still not categorical)
- When implementing the summer project, go beyond/generalize over the paper and allow weight sharing between the different embedding layers 

Why **B** makes sense:
- If I have classification or regression data with integer columns MLJFlux doesn't break
- The tutorial becomes MLJ idiomatic without coercing the indices to `Count`. They can be coerced to `Multiclass` and specify that all embedding layers (each index in the sequence treated as categorical column) should share weights.

File: docs/src/workflow examples/Composition/composition.jl (hunk 10 above)
Comment (EssamWisam) [reply docs/src/workflow examples/Composition/composition.jl:39]: Or I can remove the line and justify with a comment

**Comment by EssamWisam** [comment]
> Herculean effort, thanks!
> 
> (I haven't checked the notebooks or markdown).

Thank you so much for the high quality review. Much appreciated.

No need to check them as they are automatically generated from the Julia file. I also would quickly glance over them in the documentation to check everything is alright (e.g., the reason why I had misformed indentation for some comments is that I was trying to reverse the already misformed representation in documenter; however, that now is fixed thanks to you tab is four spaces advice).

**Comment by ablaom** [comment]
> I can see how all this should work, and it's great to see this solution, however it's definitely not MLJ idiomatic, which adds a cognitive burden for users who want to first understand the "proper" way of doing things.

I see you agree and for this reason will push back against inclusion in the manual on this crucial point. Including non-idiomatic examples in core documentation is not a good idea, in my opinion.

I do agree we could immediately expand the allowed scitype for input to include `Count` for use with genuine count data, expected to be represented as integers, which, of course, a neural network can deal with. However, your example remains non-idiomatic because we are supplying `Multiclass` data under the disguise of `Count` data (but we can then skip the coercion to and from floats).

Sorry, I'm not sure I understand the distinction in A and B, but I will say weight sharing for categorical feature encoding sounds like a substantial complexity addition from the point of view of user interface because now they have to specify exactly how that looks like. And probably later layers (prescribed by the `builder`) can achieve something similar, right? But we can discuss on our next call.

**Comment by EssamWisam** [comment]
Alrighty, I see how this makes sense from a perfection perspective or by emphasizing the experience for beginners. I have hidden that tutorial accordingly. We may merge if you have seen the EarlyStopping tutorial.

Sure. We can talk about the distinction between A and B in the next call.

**Comment by ablaom** [comment]
Thanks. Merge when ready. 

Note that I have just merged #251, which has a change that likely breaks the examples: I believe you must now use optimisers from Optimisers.jl in place of the Flux.jl optimisers (which I'm guessing will ultimately be deprecated in any case). Since we will need to update the example Manifest.toml files, I propose we update the docs in a separate PR *after* the breaking release I plan soon.

**Pull request merged**